### PR TITLE
Resetting a wrong stream

### PR DESCRIPTION
**Pull request by ddragana**
If we get data on the decoder that unblocks some header decoding and one of the stream headers is malformed, neqo will reset a wrong stream(the decoder stream)

### CODE REVIEW

File: neqo-http3/src/connection_client.rs [hunk 1]
```diff
@@ -549,7 +549,18 @@ impl Http3Client {
...
+                            if let Some(r) = self.base_handler.recv_streams.get(&stream_id) {
+                                matches!(
+                                    r.stream_type(),
+                                    Http3StreamType::Control
+                                        | Http3StreamType::Encoder
+                                        | Http3StreamType::Decoder
+                                )
+                            } else {
+                                false
+                            };
```
Comment (martinthomson) [inline neqo-http3/src/connection_client.rs:562]: You might consider moving this to a `stream_is_critical(StreamId)` function on the base handler.

File: neqo-http3/src/connection_client.rs [hunk 2]
```diff
@@ -6521,4 +6532,49 @@ mod tests {
...
+
+        // Send the encoder instructions, but delay them so that the stream is blocked on decoding headers.
+        let encoder_inst_pkt = server.conn.process(None, now());
+
+        // Send response
+        let mut d = Encoder::default();
+        hframe.encode(&mut d);
+        let d_frame = HFrame::Data { len: 3 };
+        d_frame.encode(&mut d);
+        d.encode(&[0x61, 0x62, 0x63]);
```
Comment (martinthomson) [inline neqo-http3/src/connection_client.rs:6565]: we can use b"abc" for these (shorter, easier to read)

File: neqo-http3/src/connection_client.rs [hunk 3]
```diff
@@ -6521,4 +6532,49 @@ mod tests {
...
+
+    #[test]
+    fn error_request_stream() {
+        let (mut client, mut server, request_stream_id) = connect_and_send_request(true);
+
+        setup_server_side_encoder(&mut client, &mut server);
+
+        let headers = vec![
+            Header::new(":status", "200"),
+            Header::new(":method", "GET"),
```
Comment (martinthomson) [inline neqo-http3/src/connection_client.rs:6544]:
```suggestion
            Header::new(":method", "GET"), // <- invalid
```

**Review by martinthomson — COMMENTED** [review]
LGTM.  Some suggestions, though nothing major.

File: neqo-http3/src/connection.rs [hunk 4]
```diff
@@ -224,6 +224,32 @@ impl Http3Connection {
...
+    ) -> Res<()> {
+        for stream_id in unblocked_streams {
+            qdebug!([self], "Stream {} is unblocked", stream_id);
+            if let Some(r) = self.recv_streams.get_mut(&stream_id) {
+                if let Err(e) = r
+                    .http_stream()
+                    .ok_or(Error::HttpInternal(10))?
+                    .header_unblocked(conn)
+                {
+                    if e.stream_reset_error() {
```
[the remark this answers is not in the document]
Comment (ddragana) [reply neqo-http3/src/connection.rs:240]: in one case it takes stream?_id in other case there is a RecvStream already. I have decided to leave it separate.

**Pull request merged**